### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by YaroslavOkrema**
image
image

### CODE REVIEW

File: src/App.js [hunk 1]
```diff
@@ -1,10 +1,19 @@
...
-      
-    </div>
-  );
+    return (
+        <div>
+            <Layout/>
+            <Routes>
+                <Route path='main' element={<Main/>}/>
+                <Route path='widgets' element={<Widgets/>}/>
+            </Routes>
```
Comment (Roman15122022) [inline src/App.js]: separate to another component

File: src/components/Layout.jsx [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+    const [isDrawerOpen, setDrawerOpen] = useState(false);
+
+    const toggleDrawer = () =>  {
+        setDrawerOpen(prevState => !prevState)
+    }
+
+    const handleClick = (path) => {
+        navigate(path);
+    }
```
Comment (Roman15122022) [inline src/components/Layout.jsx]: dont write logic and jsx in one file, write custom hooks for all component

File: src/components/Main.jsx [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+import {useLocation} from "../hooks/useLocation";
+import {useWeather} from "../hooks/useWeather";
+import "../styles/Main.css";
+import {useIcon} from "../hooks/useIcon";
+
+const Main = () => {
+    const location = useLocation();
+    const weatherData = useWeather(location.lat, location.lon);
+    const weatherIcon = useIcon(weatherData);
+
```
Comment (Roman15122022) [inline src/components/Main.jsx]: same

File: src/hooks/useIcon.jsx [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+    function getIcon() {
+       if (weatherData && weatherData.weather && weatherData.weather.length > 0) {
+           const iconPath = weatherData.weather[0].icon;
+           const iconUrl = `https://openweathermap.org/img/wn/${iconPath}@2x.png`;
+           setIcon(iconUrl);
+       }
+    }
+
+    return icon;
```
Comment (Roman15122022) [inline src/hooks/useIcon.jsx]: We always return obj from custom hooks return { icon }

see below

File: src/hooks/useLocation.jsx [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+        navigator.geolocation.getCurrentPosition(position => {
+            const lat = position.coords.latitude;
+            const lon = position.coords.longitude;
+
+            setLocation({ lat, lon });
+        });
+    }
+
+    return location;
+}
```
Comment (Roman15122022) [inline src/hooks/useLocation.jsx]: return { location }

File: src/hooks/useWeather.jsx [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+    async function fetchData(lat, lon) {
+        const apiKey = '642c1ad128fd816bcb4ea30d07064974';
+
+        const response = await axios.get(`https://api.openweathermap.org/data/2.5/weather?lat=${lat}&lon=${lon}&appid=${apiKey}&units=metric`);
+        setWeatherData(response.data);
+        console.log(response.data);
+    }
+
+    return weatherData;
+}
```
Comment (Roman15122022) [inline src/hooks/useWeather.jsx]: create 1 custom hook for work with requests.
That will pull up all the information by city name.
`
return {lcon, name: weather.name, humidity: weatherData.main.humidity  } 
etc
`


And another one custom hook to get name city by coords

File: src/hooks/useWeather.jsx [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+    const [weatherData, setWeatherData] = useState(null);
+
+    useEffect(() => {
+        if (lat && lon) {
+            fetchData(lat, lon);
+        }
+    }, [lat, lon]);
+
+    async function fetchData(lat, lon) {
+        const apiKey = '642c1ad128fd816bcb4ea30d07064974';
```
Comment (Roman15122022) [inline src/hooks/useWeather.jsx]: apikey => .env (google about envirment)

**Review by Roman15122022 — CHANGES_REQUESTED** [review]

File: src/hooks/useIcon.jsx [hunk 8]
```diff
@@ -0,0 +1,19 @@
+import {useEffect, useState} from "react";
+
+export function useIcon(weatherData) {
+    const [icon, setIcon] = useState(null);
+
+    useEffect(() => {
+        getIcon();
+    }, [weatherData]);
+
```
Comment (Roman15122022) [inline src/hooks/useIcon.jsx]: Its redundant

just `return { icon: getIcong() }`

File: src/hooks/useLocation.jsx [hunk 9]
```diff
@@ -0,0 +1,20 @@
+import {useEffect, useState} from "react";
+
+export function useLocation() {
+    const [location, setLocation] = useState({});
+
+    useEffect(() => {
+        getLocation();
+    }, []);
```
Comment (Roman15122022) [inline src/hooks/useLocation.jsx:8]: if` navigator.geolocation.getCurrentPosition` not async its redundant as well

File: src/hooks/useLocation.jsx [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+
+    function getLocation() {
+        navigator.geolocation.getCurrentPosition(position => {
+            const lat = position.coords.latitude;
```
Comment (Nikemaks) [inline src/hooks/useLocation.jsx]: use de structurization ex:
```
({coords}) => {
  const lat = coords.latitude ....
}
```

File: src/components/Main.jsx [hunk 11]
```diff
@@ -0,0 +1,45 @@
+import React from 'react';
+import {useLocation} from "../hooks/useLocation";
+import {useWeather} from "../hooks/useWeather";
+import "../styles/Main.css";
+import {useIcon} from "../hooks/useIcon";
+
+const Main = () => {
+    const location = useLocation();
+    const weatherData = useWeather(location.lat, location.lon);
```
Comment (Nikemaks) [inline src/components/Main.jsx]: de structurization

File: src/hooks/useLocation.jsx [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+
+export function useLocation() {
+    const [location, setLocation] = useState({});
+
+    useEffect(() => {
+        getLocation();
+    }, []);
+
+    function getLocation() {
+        navigator.geolocation.getCurrentPosition(position => {
```
Comment (Nikemaks) [inline src/hooks/useLocation.jsx]: what do you do if the user declines permission for navigation?

Comment (Roman15122022) [reply src/hooks/useLocation.jsx]: nice question)

File: src/hooks/useWeather.jsx [hunk 13]
```diff
@@ -11,9 +11,10 @@ export function useWeather(lat, lon) {
     }, [lat, lon]);
 
     async function fetchData(lat, lon) {
-        const apiKey = '642c1ad128fd816bcb4ea30d07064974';
+        const apiKey = process.env.REACT_APP_API_KEY;
+        const apiUrl = `${process.env.REACT_APP_API_URL}?lat=${lat}&lon=${lon}&appid=${apiKey}&units=metric`;
```
Comment (Roman15122022) [inline src/hooks/useWeather.jsx]: create const folder (global) and export 

`export const {
  API_KEY,
  API_URL
} = Config`

**Review by Roman15122022 — CHANGES_REQUESTED** [review]

File: src/hooks/useLocation.jsx (hunk 9 above)
Comment (YaroslavOkrema) [reply src/hooks/useLocation.jsx:8]: navigator.geolocation.getCurrentPosition async

File: src/constants/const.ts [hunk 14]
```diff
@@ -0,0 +1,2 @@
+export const API_KEY = process.env.REACT_APP_API_KEY;
+export const API_URL = process.env.REACT_APP_API_URL;
```
Comment (Roman15122022) [inline src/constants/const.ts]: if in your folder only one file - always give name indeex.ts or index,jsx

Comment (Roman15122022) [reply src/constants/const.ts]: in folders such as : helpers, constants types (global folders) always need file with name index.ts where will be imports
image
image

File: src/hooks/useWeather.jsx [hunk 15]
```diff
@@ -0,0 +1,36 @@
+import axios from "axios";
+import {useEffect, useState} from "react";
+import {API_KEY, API_URL} from "../constants/const.ts";
+
+export function useWeather(lat, lon) {
+    const [weatherData, setWeatherData] = useState(null);
+    const [icon, setIcon] = useState(null);
```
Comment (Roman15122022) [inline src/hooks/useWeather.jsx]: Rarely use null, 
if you are waiting for an object, specify `{}`
if some string then `''`
number =` 0 (or other initial)`
etc.

File: src/routes/RoutesComponent.jsx [hunk 16]
```diff
@@ -0,0 +1,17 @@
...
+import {Route, Routes} from "react-router-dom";
+import Main from "../components/Main";
+import Widgets from "../components/Widgets";
+
+const RoutesComponent = () => {
+    return (
+        <div>
+            <Routes>
+                <Route path='main' element={<Main/>}/>
+                <Route path='widgets' element={<Widgets/>}/>
```
Comment (Roman15122022) [inline src/routes/RoutesComponent.jsx]: if you not use TS then create global const for path and use it
like
```
export const MAIN = 'main'
export const WIDGETS = 'widgets'
```

File: src/hooks/useWeather.jsx [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+        const apiUrl = `${API_URL}?lat=${lat}&lon=${lon}&appid=${API_KEY}&units=metric`;
+        const response = await axios.get(apiUrl);
+        setWeatherData(response.data);
+
+        if (response.data.weather && response.data.weather.length > 0) {
+            const iconPath = response.data.weather[0].icon;
+            const iconUrl = `https://openweathermap.org/img/wn/${iconPath}@2x.png`;
+            setIcon(iconUrl);
+        }
+    }
```
Comment (Roman15122022) [inline src/hooks/useWeather.jsx]: create separate folder for fetch (get, post) requests
e.g
API => requestByCoord => index.ts
API => requestByName => index.ts

`export const geWeatherByCoords = async () => {
  return axios.get(apiUrl)
}`

**Review by Roman15122022 — CHANGES_REQUESTED** [review]

File: src/components/Main.jsx [hunk 18]
```diff
@@ -2,39 +2,45 @@ import React from 'react';
...
+    const { location} = useLocation();
+    const {
+        icon,
+        temp,
+        description,
+        humidity,
+        wind,
+        country,
+        name
+    } = useWeather(location.lat, location.lon);
```
Comment (Roman15122022) [inline src/components/Main.jsx]: use hook useLocation inside useWeather

File: src/API/requestByCoord/index.jsx [hunk 19]
```diff
@@ -0,0 +1,8 @@
+import {API_KEY, API_URL} from "../../constants";
+import axios from "axios";
+
+export async function getWeatherByCoords(lat, lon) {
+    const apiUrl = `${API_URL}?lat=${lat}&lon=${lon}&appid=${API_KEY}&units=metric`;
+    const response = await axios.get(apiUrl);
+    return response;
+}
```
Comment (Roman15122022) [inline src/API/requestByCoord/index.jsx]: index.js
If code dont return jsx dont name file .jsx

File: src/constants/routes.jsx [hunk 20]
```diff
@@ -0,0 +1,2 @@
+export const MAIN = 'main'
+export const WIDGETS = 'widgets'
```
Comment (Roman15122022) [inline src/constants/routes.jsx]: just .js

**Review by Roman15122022 — COMMENTED** [review]
Nice!

**Pull request merged**